### DISCUSSION TITLE
Btrfs Maintneance

### DISCUSSION THREAD

**Discussion by straffer44** [SpiralLinux/SpiralLinux-project#222] [General]
Just started using SprialLinux and I have never used the btrfs file system before.  For a default install on a single ssd is there any maintenance needed ?  A google search shows some believe it is not necessary on newer kernels and improvements over the years to btfrs.  

Trim is enabled in debian for ssd (fstrim.timer).

`sudo btrfs scrub start /`   could be used weekly but on a single ssd would it be useful?

`sudo btrfs balance start -musage=50 -dusage=50 /`   could be used monthly but once again useful on a single ssd?

Finally, defrag is not recommed for ssd?

Thanks for a great distro and any help

**Comment by geckolinux** [comment]
Hi there! Unfortunately there doesn't seem to be a straight answer, as there are varied opinions on the topic and it also depends on what sort of disk you have (SSD or spinning).

We had a very good thread here on this same topic, but for some weird reason it is missing. I can post the response from @Scyt4le that I have in my email notifications:

--------------------------------------

Greetings!

From openSUSE I'm familiar with some systemctl timers related to btrfs (scrub/balance if I remember correctly).
In Spiral Linux none of these are running and btrfsmaintenance packet is not installed.

Do I (or all users with btrfs and snapper) need to take action and install said packets and enable the scrub and balance services?

Scyt4le

-----------------------

Me personally am not comfortable without btrfs balance and scrup timers. What speaks against copying the defaults from openSUSE and tune it with maybe a bit less frequent because is even more conservative with patching than Leap.

What I understand btrfs maintenance helps prevent:

Over time btrfs may allocates blocks of data in a way that it cannot use some of the unused disk space. This way the system shows more free space as it can effectively use and may cause the system to stop as the system partition is full and read only. In that state btrfs balance has not enough space to reallocate the blocks to free up space. For a normal user this means the system is lost and must be reinstalled. The balance job does not prevent the system from potentially failing if the user fills up the root partition by himself but it would remove the discrepancy between displayed free disk space and usable free disk space. There are some perfomance cost when the job runs, but it is not noticeable for me.

What scrub does is not totally clear to me. I understand it in a two disk RAID1 setup, but I'm not sure if scrub does anything else as forcing the root partition to fall in an read only state after it detects errors (bitrot) when checksums doesn't fit the data. A normal end user might not understand the situation. But no scrub is just as "dangerous" as ext4, so it might be ok to not use it on the desktop. On a server or some kind of RAID it would be a crime to not use and throw this capabilities away.

Imho the current defaults are enough for Debian stable. There is not enough going on regarding updates and a lot of snapper snapshots. For testing and unstable I would find at least the balance job a very good idea. Or it's a good idea to inform the users in a wiki or before downloading that a system with the recommended btrfs disk layout performs the best with most reliability if at least 1/4 (debatable) of the available disk space remains unused.

https://github.com/kdave/btrfsmaintenance/blob/master/README.md

**Comment by straffer44** [comment]
Thanks for the information.
With the default btrfs layout in spriallinux are the commands in my first post correct ie "/" at the end?

**Reply by geckolinux** [reply]
>  With the default btrfs layout in spriallinux are the commands in my first post correct ie `/` at the end?

Correct, because it's Btrfs on root.
